test/D2L.Security.OAuth2.UnitTests/Keys/Default/KeyManagementServiceTests.cs

```diff
@@ -0,0 +1,377 @@
+			// First hour: the key ages (to allow time for caches to flush)
+			// Second hour: the key is used exclusively
+			// Third hour: we generate a new key but don't use it
+			// Forth hour: we switch to the new key
+			// Fifth hour: the old key is now expired
+			// ....
+			KeyLifetime = TimeSpan.FromHours( 4 ),
+			KeyTimeUntilUse = TimeSpan.FromHours( 1 ),
+			KeyRotationBuffer = TimeSpan.FromHours( 2 )
+		};
+
+		private Mock<IPublicKeyDataProvider> m_publicKeys;
+		private Mock<IPrivateKeyDataProvider> m_privateKeys;
+		private Mock<IDateTimeProvider> m_clock;
+
+		private IKeyManagementService m_kms;
+		private IPrivateKeyProvider m_pkp;
+		private IDisposable m_disposable;
+
+		[SetUp]
+		public void SetUp() {
+			m_publicKeys = new( MockBehavior.Strict );
+			m_privateKeys = new( MockBehavior.Strict );
+			m_clock = new( MockBehavior.Strict );
+
+			SetClock( Now );
+
+			var kms = new KeyManagementService(
+				m_publicKeys.Object,
+				m_privateKeys.Object,
+				m_clock.Object,
+				KeyConfig
+			);
+
+			m_kms = kms;
+			m_pkp = kms;
+			m_disposable = kms;
+		}
+
+		[TearDown]
+		public void TearDown() {
+			m_disposable.Dispose();
+		}
+
+		[Test]
+		public async Task GetSigningCredentialsAsync_OnBoot_NewKey_UsesIt() {
+			var key = CreateNewKey();
+
+			UsePrivateKeys( key );
+
+			using var creds = await m_pkp.GetSigningCredentialsAsync();
+
+			AssertCorrectKey( key, creds );
+		}
+
+		[Test]
+		public async Task GetSigningCredentialsAsync_OnBoot_ActiveKey_UsesIt() {
+			var key = CreateActiveKey();
+
+			UsePrivateKeys( key );
+
+			using var creds = await m_pkp.GetSigningCredentialsAsync();
+
+			AssertCorrectKey( key, creds );
+		}
+
+		[Test]
+		public async Task GetSigningCredentialsAsync_OnBoot_OldKey_UsesIt() {
+			var key = CreateOldKey();
+
+			UsePrivateKeys( key );
+
+			using var creds = await m_pkp.GetSigningCredentialsAsync();
+
+			AssertCorrectKey( key, creds );
+		}
+
+		[Test]
+		public async Task GetSigningCredentialsAsync_OnBoot_ExpiredKey_ReturnsNull() {
+			var key = CreateExpiredKey();
+
+			UsePrivateKeys( key );
+
+			using var creds = await m_pkp.GetSigningCredentialsAsync();
+
+			Assert.IsNull( creds );
+		}
+
+		[Test]
+		public async Task GetSigningCredentialsAsync_HasNewKeyCached_SticksToIt() {
+			var firstKey = CreateNewKey();
+
+			await PrimeCacheAsync( firstKey );
+
+			// It would be "better" to use this, but GetSigningCredentialsAsync()
+			// won't go looking. RefreshAsync() is used to do that out-of-band.
+			var secondKey = CreateActiveKey();
+
+			UsePrivateKeys( firstKey, secondKey );
+
+			using var creds = await m_pkp.GetSigningCredentialsAsync();
+
+			AssertCorrectKey( firstKey, creds );
+		}
+
+		[Test]
+		public async Task GetSigningCredentialsAsync_TwoActiveKeys_PrefersNewest() {
+			var oldActive = CreateActiveKey();
+			var newActive = CreateNewActiveKey();
+
+			UsePrivateKeys( oldActive, newActive );
+
+			await m_kms.RefreshKeyAsync();

```

Comment:
nit: I think this test aligns more with the `OnBoot` tests, allowing `GetSigningCredentialsAsync` to call `RefreshAsync` itself, given we're going through it?